<a href="https://colab.research.google.com/github/Blackman9t/PandasCookbook/blob/master/Chapter_3_Beginning_Data_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Chapter 3: Beginning Data Analysis

## Recipes
* [Developing a data analysis routine](#Developing-a-data-analysis-routine)
* [Reducing memory by changing data types](#Reducing-memory-by-changing-data-types)
* [Selecting the smallest of the largest](#Selecting-the-smallest-of-the-largest)
* [Selecting the largest of each group by sorting](#Selecting-the-largest-of-each-group-by-sorting)
* [Duplicating nlargest with sort_values](#Duplicating-nlargest-with-sort_values)
* [Calculating a trailing stop order price](#Calculating-a-trailing-stop-order-price)

In [0]:
import pandas as pd
import numpy as np
pd.options.display.max_columns = 50

# Developing a data analysis routine

It is important to consider the steps that you, as an analyst, take when you first encounter a dataset after importing it into your workspace as a DataFrame. Is there a set of tasks that you usually undertake to first examine the data? Are you aware of all the possible data types?

Although there is no standard approach when beginning a data analysis, it is typically a good idea to develop a routine for yourself when first examining a dataset. Similar to common routines that we have for waking up, showering, going to work, eating, and so on, a beginning data analysis routine helps one quickly get acquainted with a new dataset. This routine can manifest itself as a dynamic checklist of tasks that evolves as your familiarity with pandas and data analysis expands.

Exploratory Data Analysis (EDA) is a term used to encompass the entire process of analyzing data without the formal use of statistical testing procedures. Much of EDA involves visually displaying different relationships among the data to detect interesting patterns and develop hypotheses.

**Getting ready**<br>
This recipe covers a small but fundamental part of EDA: the collection of metadata and univariate descriptive statistics in a routine and systematic way. It outlines a common set of tasks that can be undertaken when first importing any dataset as a pandas DataFrame. This recipe may help form the basis of the routine that you can implement when first examining a dataset.

<font color='khaki'>Note<br>
Metadata describes the dataset, or more aptly, data about the data. Examples of metadata include the number of columns/rows, column names, data types of each column, the source of the dataset, the date of collection, the acceptable values for different columns, and so on. Univariate descriptive statistics are summary statistics about individual variables (columns) of the dataset, independent of all other variables.</font>

How to do it...<br>
**Step One:**<br>
First, some metadata on the college dataset will be collected, followed by basic summary statistics of each column:

Read in the dataset, and view the first five rows with the head method:

In [0]:
college_raw = 'https://raw.githubusercontent.com/Blackman9t/PandasCookbook/master/college.csv'  # raw link from github
college = pd.read_csv(college_raw)

In [3]:
college.head()

INSTNM        CITY STABBR  HBCU  MENONLY  \
0             Alabama A & M University      Normal     AL   1.0      0.0   
1  University of Alabama at Birmingham  Birmingham     AL   0.0      0.0   
2                   Amridge University  Montgomery     AL   0.0      0.0   
3  University of Alabama in Huntsville  Huntsville     AL   0.0      0.0   
4             Alabama State University  Montgomery     AL   1.0      0.0   

   WOMENONLY  RELAFFIL  SATVRMID  SATMTMID  DISTANCEONLY     UGDS  UGDS_WHITE  \
0        0.0         0     424.0     420.0           0.0   4206.0      0.0333   
1        0.0         0     570.0     565.0           0.0  11383.0      0.5922   
2        0.0         1       NaN       NaN           1.0    291.0      0.2990   
3        0.0         0     595.0     590.0           0.0   5451.0      0.6988   
4        0.0         0     425.0     430.0           0.0   4811.0      0.0158   

   UGDS_BLACK  UGDS_HISP  UGDS_ASIAN  UGDS_AIAN  UGDS_NHPI  UGDS_2MOR  \
0      0.9353     0.0055      0.0019     0.0024     0.0019     0.0000   
1      0.2600     0.0283      0.0518     0.0022     0.0007     0.0368   
2      0.4192     0.0069      0.0034     0.0000     0.0000     0.0000   
3      0.1255     0.0382      0.0376     0.0143     0.0002     0.0172   
4      0.9208     0.0121      0.0019     0.0010     0.0006     0.0098   

   UGDS_NRA  UGDS_UNKN  PPTUG_EF  CURROPER  PCTPELL  PCTFLOAN  UG25ABV  \
0    0.0059     0.0138    0.0656         1   0.7356    0.8284   0.1049   
1    0.0179     0.0100    0.2607         1   0.3460    0.5214   0.2422   
2    0.0000     0.2715    0.4536         1   0.6801    0.7795   0.8540   
3    0.0332     0.0350    0.2146         1   0.3072    0.4596   0.2640   
4    0.0243     0.0137    0.0892         1   0.7347    0.7554   0.1270   

  MD_EARN_WNE_P10 GRAD_DEBT_MDN_SUPP  
0           30300              33888  
1           39700            21941.5  
2           40100              23370  
3           45500              24097  
4           26600            33118.5

**Step two**

Get the dimensions of the DataFrame with the shape attribute:


In [4]:
college.shape

(7535, 27)

**Step three**:

List the data type of each column, number of non-missing values, and memory usage with the info method:

In [6]:
college.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7535 entries, 0 to 7534
Data columns (total 27 columns):
INSTNM                7535 non-null object
CITY                  7535 non-null object
STABBR                7535 non-null object
HBCU                  7164 non-null float64
MENONLY               7164 non-null float64
WOMENONLY             7164 non-null float64
RELAFFIL              7535 non-null int64
SATVRMID              1185 non-null float64
SATMTMID              1196 non-null float64
DISTANCEONLY          7164 non-null float64
UGDS                  6874 non-null float64
UGDS_WHITE            6874 non-null float64
UGDS_BLACK            6874 non-null float64
UGDS_HISP             6874 non-null float64
UGDS_ASIAN            6874 non-null float64
UGDS_AIAN             6874 non-null float64
UGDS_NHPI             6874 non-null float64
UGDS_2MOR             6874 non-null float64
UGDS_NRA              6874 non-null float64
UGDS_UNKN             6874 non-null float64
PPTUG_EF          

**Step four**

Get summary statistics for the numerical columns and transpose the DataFrame for more readable output:


In [7]:
with pd.option_context('display.max_rows', 8):
    display(college.describe(include=[np.number]).T)  # displaying max 8 rows, and transposing the output for readability

count      mean       std  min     25%      50%       75%  max
HBCU       7164.0  0.014238  0.118478  0.0  0.0000  0.00000  0.000000  1.0
MENONLY    7164.0  0.009213  0.095546  0.0  0.0000  0.00000  0.000000  1.0
WOMENONLY  7164.0  0.005304  0.072642  0.0  0.0000  0.00000  0.000000  1.0
RELAFFIL   7535.0  0.190975  0.393096  0.0  0.0000  0.00000  0.000000  1.0
...           ...       ...       ...  ...     ...      ...       ...  ...
CURROPER   7535.0  0.923291  0.266146  0.0  1.0000  1.00000  1.000000  1.0
PCTPELL    6849.0  0.530643  0.225544  0.0  0.3578  0.52150  0.712900  1.0
PCTFLOAN   6849.0  0.522211  0.283616  0.0  0.3329  0.58330  0.745000  1.0
UG25ABV    6718.0  0.410021  0.228939  0.0  0.2415  0.40075  0.572275  1.0

[22 rows x 8 columns]

**Step five**

Get summary statistics for the object and categorical columns:


Categorical columns are usually going to be either of type np.object or pd.Categorical. Step 5 ensures that both of these types are represented. In both step 4 and step 5, the output DataFrame is transposed with the T attribute. This eases readability for DataFrames with many columns.

In [13]:
college.describe(include=[np.object, pd.Categorical]).T

count unique                                top  freq
INSTNM              7535   7535  Regency Beauty Institute-Rockford     1
CITY                7535   2514                           New York    87
STABBR              7535     59                                 CA   773
MD_EARN_WNE_P10     6413    598                  PrivacySuppressed   822
GRAD_DEBT_MDN_SUPP  7503   2038                  PrivacySuppressed  1510

How it works...
After importing your dataset, a common task is to print out the first few rows of the DataFrame for manual inspection with the head method. The shape attribute returns the first piece of metadata, a tuple containing the number of rows and columns.

The primary method to get the most metadata at once is the info method. It provides each column name, the number of non-missing values, the data type of each column, and the approximate memory usage of the DataFrame. For all DataFrames, columns values are always one data type. The same holds for relational databases. DataFrames, as a whole, might be composed of columns with different data types.

**See numpy Data type Hierarchy below**

<img src='https://github.com/Blackman9t/PandasCookbook/blob/master/numpydtypehierarchy.png?raw=true' height=400>

<font color='khaki'>Note<br>
Broadly speaking, we can classify data as being either continuous or categorical. Continuous data is always numeric and can usually take on an infinite number of possibilities such as height, weight, and salary. Categorical data represents discrete values that take on a finite number of possibilities such as ethnicity, employment status, and car color. Categorical data can be represented numerically or with characters.</font>

There's more...
It is possible to specify the exact quantiles returned from the describe method when used with numeric columns:

In [0]:
# used to shorten info output.
import io
buf = io.StringIO()
college.info(buf=buf)
s = buf.getvalue()

In [0]:
s

"<class 'pandas.core.frame.DataFrame'>\nRangeIndex: 7535 entries, 0 to 7534\nData columns (total 27 columns):\nINSTNM                7535 non-null object\nCITY                  7535 non-null object\nSTABBR                7535 non-null object\nHBCU                  7164 non-null float64\nMENONLY               7164 non-null float64\nWOMENONLY             7164 non-null float64\nRELAFFIL              7535 non-null int64\nSATVRMID              1185 non-null float64\nSATMTMID              1196 non-null float64\nDISTANCEONLY          7164 non-null float64\nUGDS                  6874 non-null float64\nUGDS_WHITE            6874 non-null float64\nUGDS_BLACK            6874 non-null float64\nUGDS_HISP             6874 non-null float64\nUGDS_ASIAN            6874 non-null float64\nUGDS_AIAN             6874 non-null float64\nUGDS_NHPI             6874 non-null float64\nUGDS_2MOR             6874 non-null float64\nUGDS_NRA              6874 non-null float64\nUGDS_UNKN             6874 non-null flo

In [0]:
s1 = "<class 'pandas.core.frame.DataFrame'>\nRangeIndex: 7535 entries, 0 to 7534\nData columns (total 27 columns):\nINSTNM                7535 non-null object\nCITY                  7535 non-null object\nSTABBR                7535 non-null object\nHBCU                  7164 non-null float64\n..."

In [0]:
s2 = "\nPCTFLOAN              6849 non-null float64\nUG25ABV               6718 non-null float64\nMD_EARN_WNE_P10       6413 non-null object\nGRAD_DEBT_MDN_SUPP    7503 non-null object\ndtypes: float64(20), int64(2), object(5)\nmemory usage: 1.6+ MB\n"

In [17]:
print(s1 + s2)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7535 entries, 0 to 7534
Data columns (total 27 columns):
INSTNM                7535 non-null object
CITY                  7535 non-null object
STABBR                7535 non-null object
HBCU                  7164 non-null float64
...
PCTFLOAN              6849 non-null float64
UG25ABV               6718 non-null float64
MD_EARN_WNE_P10       6413 non-null object
GRAD_DEBT_MDN_SUPP    7503 non-null object
dtypes: float64(20), int64(2), object(5)
memory usage: 1.6+ MB



In [18]:
college.describe(include=[np.number]).T

count      mean       std  min     25%      50%       75%  max
HBCU       7164.0  0.014238  0.118478  0.0  0.0000  0.00000  0.000000  1.0
MENONLY    7164.0  0.009213  0.095546  0.0  0.0000  0.00000  0.000000  1.0
WOMENONLY  7164.0  0.005304  0.072642  0.0  0.0000  0.00000  0.000000  1.0
RELAFFIL   7535.0  0.190975  0.393096  0.0  0.0000  0.00000  0.000000  1.0
...           ...       ...       ...  ...     ...      ...       ...  ...
CURROPER   7535.0  0.923291  0.266146  0.0  1.0000  1.00000  1.000000  1.0
PCTPELL    6849.0  0.530643  0.225544  0.0  0.3578  0.52150  0.712900  1.0
PCTFLOAN   6849.0  0.522211  0.283616  0.0  0.3329  0.58330  0.745000  1.0
UG25ABV    6718.0  0.410021  0.228939  0.0  0.2415  0.40075  0.572275  1.0

[22 rows x 8 columns]

In [19]:
college.describe(include=[np.object, pd.Categorical]).T

count unique                                top  freq
INSTNM              7535   7535  Regency Beauty Institute-Rockford     1
CITY                7535   2514                           New York    87
STABBR              7535     59                                 CA   773
MD_EARN_WNE_P10     6413    598                  PrivacySuppressed   822
GRAD_DEBT_MDN_SUPP  7503   2038                  PrivacySuppressed  1510

## There's more...

Code below prints out the describe statistics for numerical columns and includes the percentile param that sets fixed pct to be displayed too

In [21]:
with pd.option_context('display.max_rows', 5):
    display(college.describe(include=[np.number], 
                 percentiles=[.01, .05, .10, .25, .5, .75, .9, .95, .99]).T)

count      mean       std  min      1%      5%     10%     25%  \
HBCU      7164.0  0.014238  0.118478  0.0  0.0000  0.0000  0.0000  0.0000   
MENONLY   7164.0  0.009213  0.095546  0.0  0.0000  0.0000  0.0000  0.0000   
...          ...       ...       ...  ...     ...     ...     ...     ...   
PCTFLOAN  6849.0  0.522211  0.283616  0.0  0.0000  0.0000  0.0000  0.3329   
UG25ABV   6718.0  0.410021  0.228939  0.0  0.0025  0.0374  0.0899  0.2415   

              50%       75%      90%      95%       99%  max  
HBCU      0.00000  0.000000  0.00000  0.00000  1.000000  1.0  
MENONLY   0.00000  0.000000  0.00000  0.00000  0.000000  1.0  
...           ...       ...      ...      ...       ...  ...  
PCTFLOAN  0.58330  0.745000  0.84752  0.89792  0.986368  1.0  
UG25ABV   0.40075  0.572275  0.72666  0.80000  0.917383  1.0  

[22 rows x 14 columns]

# Data Dictionary

A crucial part of a data analysis involves creating and maintaining a data dictionary. A data dictionary is a table of metadata and notes on each column of data. One of the primary purposes of a data dictionary is to explain the meaning of the column names. The college dataset uses a lot of abbreviations that are likely to be unfamiliar to an analyst who is inspecting it for the first time.

A data dictionary for the college dataset is provided in the following college_data_dictionary.csv file:



In [0]:
data_dict_raw = 'https://raw.githubusercontent.com/Blackman9t/PandasCookbook/master/college_data_dictionary.csv'
college_dd = pd.read_csv(data_dict_raw)

In [24]:
with pd.option_context('display.max_rows', 10):
    display(college_dd)

column_name                                description
0               INSTNM                           Institution Name
1                 CITY                              City Location
2               STABBR                         State Abbreviation
3                 HBCU   Historically Black College or University
4              MENONLY                               0/1 Men Only
..                 ...                                        ...
22             PCTPELL           Percent Students with Pell grant
23            PCTFLOAN         Percent Students with federal loan
24             UG25ABV             Percent Students Older than 25
25     MD_EARN_WNE_P10  Median Earnings 10 years after enrollment
26  GRAD_DEBT_MDN_SUPP                  Median debt of completers

[27 rows x 2 columns]

As you can see, it is immensely helpful in deciphering the abbreviated column names. DataFrames are actually not the best place to store data dictionaries. A platform such as Excel or Google Sheets with easy ability to edit values and append columns is a better choice. Minimally, a column to keep track of notes on the data should be included in a data dictionary. A data dictionary is one of the first things that you can share as an analyst to collaborators.

It will often be the case that the dataset you are working with originated from a database whose administrators you will have to contact in order to get more information. Formal electronic databases generally have more formal representations of their data, called schemas. If possible, attempt to investigate your dataset with people who have expert knowledge on its design.

# Reducing memory by changing data types

Pandas does not broadly classify data as either continuous or categorical but has precise technical definitions for many distinct data types.

Getting ready
This recipe changes the data type of one of the object columns from the college dataset to the special pandas Categorical data type to drastically reduce its memory usage.

How to do it...
After reading in our college dataset, we select a few columns of different data types that will clearly show how much memory may be saved:

In [56]:
different_cols = ['RELAFFIL', 'SATMTMID', 'CURROPER', 'INSTNM', 'STABBR']
college[different_cols].dtypes

col2 = college.loc[:, different_cols]
col2.head()

RELAFFIL  SATMTMID  CURROPER                               INSTNM STABBR
0         0     420.0         1             Alabama A & M University     AL
1         0     565.0         1  University of Alabama at Birmingham     AL
2         1       NaN         1                   Amridge University     AL
3         0     590.0         1  University of Alabama in Huntsville     AL
4         0     430.0         1             Alabama State University     AL

Let's check the original memory of col2

In [57]:
original_mem = col2.memory_usage(deep=True)
original_mem

Index           80
RELAFFIL      7535
SATMTMID     60280
CURROPER     60280
INSTNM      660699
STABBR      444565
dtype: int64

In [0]:
college['RELAFFIL'] = college['RELAFFIL'].astype(np.int8)

Inspect the data types of each column:

In [58]:
col2.dtypes

RELAFFIL       int8
SATMTMID    float64
CURROPER      int64
INSTNM       object
STABBR       object
dtype: object

Find the memory usage of each column with the memory_usage method:

In [59]:
col2[different_cols].memory_usage(deep=True)

Index           80
RELAFFIL      7535
SATMTMID     60280
CURROPER     60280
INSTNM      660699
STABBR      444565
dtype: int64

There is no need to use 64 bits for the CURROPER column as it contains only 0/1 values. Let's convert this column to an 8-bit (1 byte) integer with the astype method:

In [0]:
col2.CURROPER = col2.CURROPER.astype('int8')

Use the dtypes attribute to confirm the data type change:


In [61]:
col2.dtypes

RELAFFIL       int8
SATMTMID    float64
CURROPER       int8
INSTNM       object
STABBR       object
dtype: object

To save even more memory, you will want to consider changing object data types to categorical if they have a reasonably low cardinality (number of unique values). Let's first check the number of unique values for both the object columns:

In [62]:
col2.select_dtypes(include=['object']).nunique()  # selecting object data types and checking count of unique elements using nunique() param

INSTNM    7535
STABBR      59
dtype: int64

The STABBR column is a good candidate to convert to Categorical as less than one percent of its values are unique:

In [63]:
col2.STABBR = col2.STABBR.astype('category')

col2.dtypes

RELAFFIL        int8
SATMTMID     float64
CURROPER        int8
INSTNM        object
STABBR      category
dtype: object

Compute the memory usage again:

In [64]:
new_mem = col2.memory_usage(deep=True)
new_mem

Index           80
RELAFFIL      7535
SATMTMID     60280
CURROPER      7535
INSTNM      660699
STABBR       13576
dtype: int64

Finally, let's compare the original memory usage with our updated memory usage. The RELAFFIL column is, as expected, an eighth of its original, while the STABBR column has shrunk to just three percent of its original size:


In [65]:
pct_mem = new_mem / original_mem

pct_mem

Index       1.000000
RELAFFIL    1.000000
SATMTMID    1.000000
CURROPER    0.125000
INSTNM      1.000000
STABBR      0.030538
dtype: float64

How it works...
Pandas defaults integer and float data types to 64 bits regardless of the maximum necessary size for the particular DataFrame. Integers, floats, and even booleans may be coerced to a different data type with the astype method and passing it the exact type, either as a string or specific object, as done in step 4.

The RELAFFIL column is a good choice to cast to a smaller integer type as the data dictionary explains that its values must be 0/1. 

<font color='khaki'>Note<br>
The memory units displayed are in bytes and not bits. One byte is equivalent to 8 bits, so when RELAFFIL was changed to an 8-bit integer, it uses one 1 byte of memory and as there are 7,535 rows, its memory footprint is equivalent to 7,535 bytes.

Columns that are object data type, such as INSTNM, are not like the other pandas data types. For all the other pandas data types, each value in that column is the same data type. For instance, when a column has the int64 type, every individual column value is also int64. This is not true for columns that are object data type. Each individual column value can be of any type. Object data types can have a mix of strings, numerics, datetimes, or even other Python objects such as lists or tuples. For this reason, the object data type is sometimes referred to as a catch-all for a column of data that doesn't match any of the other data types. The vast majority of the time, though, object data type columns will all be strings.

<font color='khaki'>Note<br>
Relational database management systems such as Microsoft's SQL Server or PostgreSQL have specific data types for characters such as varchar, text, or nchar that also usually specify a maximum number of characters. Pandas object data type is a much broader data type. Every value in an object column can be of any data type.

Therefore, the memory of each individual value in an object data type column is inconsistent. There is no predefined amount of memory for each value like the other data types. For pandas to extract the exact amount of memory of an object data type column, the deep parameter must be set to True in the memory_usage method.

Object columns are targets for the largest memory savings. Pandas has an additional categorical data type that is not available in NumPy. When converting to category, pandas internally creates a mapping from integers to each unique string value. Thus, each string only needs to be kept a single time in memory. As you can see, this simple change of data type reduced memory usage by 97%.

You might also have noticed that the index uses an extremely low amount of memory. If no index is specified during DataFrame creation, as is the case in this recipe, pandas defaults the index to a RangeIndex. The RangeIndex is very similar to the built-in range function. It produces values on demand and only stores the minimum amount of information needed to create an index.

## There's more...

In [0]:
college = pd.read_csv(college_raw)

In [67]:
college[['CURROPER', 'INSTNM']].memory_usage(deep=True)

Index           80
CURROPER     60280
INSTNM      660240
dtype: int64

To get a better idea of how object data type columns differ from integers and floats, a single value from each one of these columns can be modified and the resulting memory usage displayed. The CURROPER and INSTNM columns are of int64 and object types, respectively:

In [68]:
college.loc[0, 'CURROPER'] = 10000000
college.loc[0, 'INSTNM'] = college.loc[0, 'INSTNM'] + 'a'
# college.loc[1, 'INSTNM'] = college.loc[1, 'INSTNM'] + 'a'
college[['CURROPER', 'INSTNM']].memory_usage(deep=True)

Index           80
CURROPER     60280
INSTNM      660345
dtype: int64

Memory usage for CURROPER remained the same since a 64-bit integer is more than enough space for the larger number. On the other hand, the memory usage for INSTNM increased by 105 bytes by just adding a single letter to one value.

<font color='khaki'>Note<br>
Python 3 uses Unicode, a standardized character representation intended to encode all the world's writing systems. Unicode uses up to 4 bytes per character. It seems that pandas has some overhead (100 bytes) when making the first modification to a character value. Afterward, increments of 5 bytes per character are sustained.

Not all columns can be coerced to the desired type. Take a look at the MENONLY column, which from the data dictionary appears to contain only 0/1 values. The actual data type of this column upon import unexpectedly turns out to be float64. The reason for this is that there happen to be missing values, denoted by np.nan. There is no integer representation for missing values. Any numeric column with even a single missing value must be a float. Furthermore, any column of an integer data type will automatically be coerced to a float if one of the values becomes missing:

In [69]:
college['MENONLY'].dtype

dtype('float64')

In [0]:
# college['MENONLY'].astype('int8') # ValueError: Cannot convert non-finite values (NA or inf) to integer

# ValueError: Cannot convert non-finite values (NA or inf) to integer

Additionally, it is possible to substitute string names in place of Python objects when referring to data types. For instance, when using the include parameter in the describe DataFame method, it is possible to pass a list of either the formal object NumPy/pandas object or their equivalent string representation

In [72]:
college.describe(include=['int64', 'float64']).T

count         mean            std  min     25%      50%       75%  \
HBCU       7164.0     0.014238       0.118478  0.0  0.0000  0.00000  0.000000   
MENONLY    7164.0     0.009213       0.095546  0.0  0.0000  0.00000  0.000000   
WOMENONLY  7164.0     0.005304       0.072642  0.0  0.0000  0.00000  0.000000   
RELAFFIL   7535.0     0.190975       0.393096  0.0  0.0000  0.00000  0.000000   
...           ...          ...            ...  ...     ...      ...       ...   
CURROPER   7535.0  1328.063172  115201.552429  0.0  1.0000  1.00000  1.000000   
PCTPELL    6849.0     0.530643       0.225544  0.0  0.3578  0.52150  0.712900   
PCTFLOAN   6849.0     0.522211       0.283616  0.0  0.3329  0.58330  0.745000   
UG25ABV    6718.0     0.410021       0.228939  0.0  0.2415  0.40075  0.572275   

                  max  
HBCU              1.0  
MENONLY           1.0  
WOMENONLY         1.0  
RELAFFIL          1.0  
...               ...  
CURROPER   10000000.0  
PCTPELL           1.0  
PCTFLOAN          1.0  
UG25ABV           1.0  

[22 rows x 8 columns]

In [73]:
college.describe(include=[np.int64, np.float64]).T

count         mean            std  min     25%      50%       75%  \
HBCU       7164.0     0.014238       0.118478  0.0  0.0000  0.00000  0.000000   
MENONLY    7164.0     0.009213       0.095546  0.0  0.0000  0.00000  0.000000   
WOMENONLY  7164.0     0.005304       0.072642  0.0  0.0000  0.00000  0.000000   
RELAFFIL   7535.0     0.190975       0.393096  0.0  0.0000  0.00000  0.000000   
...           ...          ...            ...  ...     ...      ...       ...   
CURROPER   7535.0  1328.063172  115201.552429  0.0  1.0000  1.00000  1.000000   
PCTPELL    6849.0     0.530643       0.225544  0.0  0.3578  0.52150  0.712900   
PCTFLOAN   6849.0     0.522211       0.283616  0.0  0.3329  0.58330  0.745000   
UG25ABV    6718.0     0.410021       0.228939  0.0  0.2415  0.40075  0.572275   

                  max  
HBCU              1.0  
MENONLY           1.0  
WOMENONLY         1.0  
RELAFFIL          1.0  
...               ...  
CURROPER   10000000.0  
PCTPELL           1.0  
PCTFLOAN          1.0  
UG25ABV           1.0  

[22 rows x 8 columns]

These strings can be similarly used when changing types:

In [0]:
college['RELAFFIL'] = college['RELAFFIL'].astype('int8')

In [77]:
college.describe(include=['int', 'float']).T  # defaults to 64 bit int/floats

count         mean            std    min       25%        50%  \
HBCU       7164.0     0.014238       0.118478    0.0    0.0000    0.00000   
MENONLY    7164.0     0.009213       0.095546    0.0    0.0000    0.00000   
WOMENONLY  7164.0     0.005304       0.072642    0.0    0.0000    0.00000   
SATVRMID   1185.0   522.819409      68.578862  290.0  475.0000  510.00000   
...           ...          ...            ...    ...       ...        ...   
CURROPER   7535.0  1328.063172  115201.552429    0.0    1.0000    1.00000   
PCTPELL    6849.0     0.530643       0.225544    0.0    0.3578    0.52150   
PCTFLOAN   6849.0     0.522211       0.283616    0.0    0.3329    0.58330   
UG25ABV    6718.0     0.410021       0.228939    0.0    0.2415    0.40075   

                  75%         max  
HBCU         0.000000         1.0  
MENONLY      0.000000         1.0  
WOMENONLY    0.000000         1.0  
SATVRMID   555.000000       765.0  
...               ...         ...  
CURROPER     1.000000  10000000.0  
PCTPELL      0.712900         1.0  
PCTFLOAN     0.745000         1.0  
UG25ABV      0.572275         1.0  

[21 rows x 8 columns]

In [78]:
college.describe(include=['number']).T  # also works as the default int/float are 64 bits

count         mean            std  min     25%      50%       75%  \
HBCU       7164.0     0.014238       0.118478  0.0  0.0000  0.00000  0.000000   
MENONLY    7164.0     0.009213       0.095546  0.0  0.0000  0.00000  0.000000   
WOMENONLY  7164.0     0.005304       0.072642  0.0  0.0000  0.00000  0.000000   
RELAFFIL   7535.0     0.190975       0.393096  0.0  0.0000  0.00000  0.000000   
...           ...          ...            ...  ...     ...      ...       ...   
CURROPER   7535.0  1328.063172  115201.552429  0.0  1.0000  1.00000  1.000000   
PCTPELL    6849.0     0.530643       0.225544  0.0  0.3578  0.52150  0.712900   
PCTFLOAN   6849.0     0.522211       0.283616  0.0  0.3329  0.58330  0.745000   
UG25ABV    6718.0     0.410021       0.228939  0.0  0.2415  0.40075  0.572275   

                  max  
HBCU              1.0  
MENONLY           1.0  
WOMENONLY         1.0  
RELAFFIL          1.0  
...               ...  
CURROPER   10000000.0  
PCTPELL           1.0  
PCTFLOAN          1.0  
UG25ABV           1.0  

[22 rows x 8 columns]

In [0]:
college['MENONLY'] = college['MENONLY'].astype('float16')
college['RELAFFIL'] = college['RELAFFIL'].astype('int8')

Lastly, it is possible to see the enormous memory difference between the minimal RangeIndex and Int64Index, which stores every row index in memory:



In [80]:
college.index = pd.Int64Index(college.index)
college.index.memory_usage()

60280

# Selecting the smallest of the largest

This recipe can be used to create catchy news headlines such as Out of the top 100 best universities, these 5 have the lowest tuition or From the top 50 cities to live, these 10 are the most affordable. During an analysis, it is possible that you will first need to find a grouping of data that contains the top n values in a single column and, from this subset, find the bottom m values based on a different column.

In [83]:
diamonds = pd.read_csv('https://raw.githubusercontent.com/Blackman9t/PandasCookbook/master/diamonds.csv')
diamonds.head()

carat      cut color clarity  depth  table  price     x     y     z
0   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
1   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31
2   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31
3   0.29  Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63
4   0.31     Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75

In [84]:
diamonds.shape

(53940, 10)

In [86]:
diamonds2 = diamonds[['cut', 'carat', 'price']]
diamonds2.head()

cut  carat  price
0    Ideal   0.23    326
1  Premium   0.21    326
2     Good   0.23    327
3  Premium   0.29    334
4     Good   0.31    335

Use the nlargest method to select the top 100 diamonds by carat:

In [87]:
diamonds2.nlargest(100,'carat').head()

cut  carat  price
27415     Fair   5.01  18018
27630     Fair   4.50  18531
27130     Fair   4.13  17329
25998  Premium   4.01  15223
25999  Premium   4.01  15223

Chain the nsmallest method to return the ten lowest diamond prices  among those with a top 100 score:

In [89]:
pd.set_option('max_row',10)
#movie2.nlargest(100, 'imdb_score').nsmallest(5, 'budget')
diamonds2.nlargest(100, 'carat').nsmallest(10,'price')

cut  carat  price
16283  Very Good   3.00   6512
17196       Fair   2.72   6870
19339    Premium   3.01   8040
19346       Fair   3.00   8044
19866    Premium   2.68   8419
20462       Fair   2.74   8807
21566  Very Good   2.68   9665
21758       Fair   3.11   9823
21862    Premium   3.01   9925
22399    Premium   2.77  10424

How it works...
The first parameter of the nlargest method, n, must be an integer and selects the number of rows to be returned. The second parameter, columns, takes a column name as a string. Step 2 returns the 100 highest carat diamonds. We could have saved this intermediate result as its own variable but instead, we chain the nsmallest method to it in step 3, which returns exactly ten rows, sorted by lowest price.

There's more...
It is possible to pass a list of column names to the columns parameter of the nlargest/nsmallest methods. This would only be useful to break ties in the event that there were duplicate values sharing the nth ranked spot in the first column in the list.

# Selecting the largest of each group by sorting

One of the most basic and common operations to perform during a data analysis is to select rows containing the largest value of some column within a group. For instance, this would be like finding the highest rated film of each year or the highest grossing film by content rating. To accomplish this task, we need to sort the groups as well as the column used to rank each member of the group, and then extract the highest member of each group.

In this recipe, we will find the depth of the highest rated diamonds of each cut category.

How to do it...

First let's confirm the categories in the cut column

In [100]:
diamonds.groupby('cut').size()

cut
Fair          1610
Good          4906
Ideal        21551
Premium      13791
Very Good    12082
dtype: int64

next let's slim the data set to just the columns we need to sort by cut category

In [107]:
diamonds2 = diamonds[['cut','depth']]

diamonds2.head()

cut  depth
0    Ideal   61.5
1  Premium   59.8
2     Good   56.9
3  Premium   62.4
4     Good   63.3

Use the sort_values method to sort the DataFrame by cut. The default behavior sorts from the smallest to largest. Use the ascending parameter to invert this behavior by setting it equal to False:

In [108]:
diamonds2.sort_values('cut', ascending=False).head(8)

cut  depth
8748   Very Good   61.7
23064  Very Good   62.4
23058  Very Good   63.3
31236  Very Good   60.2
19154  Very Good   63.2
23055  Very Good   63.0
31238  Very Good   63.0
43477  Very Good   59.4

Notice how only the cut was sorted. To sort multiple columns at once, use a list. Let's look at how to sort both cut and depth:

In [109]:
diamonds3 = diamonds2.sort_values(['cut','depth'], ascending=False)
diamonds3.head(8)

cut  depth
17100  Very Good   64.9
9160   Very Good   64.7
14572  Very Good   64.7
5726   Very Good   64.6
8089   Very Good   64.6
43526  Very Good   64.6
1159   Very Good   64.5
28286  Very Good   64.5

Now, we use the drop_duplicates method to keep only the first row of each cut category:

In [110]:
best_cut_diamonds_depth = diamonds3.drop_duplicates(subset='cut')
best_cut_diamonds_depth

cut  depth
17100  Very Good   64.9
106      Premium   63.0
26263      Ideal   66.7
15227       Good   67.0
52860       Fair   79.0

And from this brief exercise, we can see that:<br>
The highest depth of a diamond in the  Premium class is just 63.0 inches.<br>
While the highest depth of a  diamond belonging to Fair class is  79.0 inches.<br>
This means the lower the depth, the better the quality as we can see in other classes such as Very-good, Ideal, and Good.<br>
So therefore diamond ratings by cut and depth could be shown like this in order of listing<br>
Premium- 63.0<br>
Very Good- 64.9<br>
Ideal- 66.7<br>
Good- 67.0<br>
Fair- 79.0<br>

How it works...
In step 1, we slim the dataset down to concentrate on only the columns of importance. This recipe would work the same with the entire DataFrame. <br>
Step 2 shows how to sort a DataFrame by a single column, which is not exactly what we wanted. <br>
Step 3 sorts multiple columns at the same time. It works by first sorting all of cut and then, within each distinct value of cut,  it sorts by price.

The default behavior of the drop_duplicates method is to keep the first occurrence of each unique row, which would not drop any rows as each row is unique. However, the subset parameter alters it to only consider the column (or list of columns) given to it. <br>
In this example, only one row for each cut  will be returned. As we sorted by cut and price in the last step, the highest scoring price for each cut is what we get.

###There's more...

It is possible to sort one column in ascending order while simultaneously sorting another column in descending order. To accomplish this, pass in a list of booleans to the ascending parameter that corresponds to how you would like each column sorted. The following sorts cut and carat in descending order and price in ascending order. It then finds the cheapest diamond for each cut and carat group:



In [112]:
#  First group the columns we want to a data frame variable
diamonds4 = diamonds[['cut','carat','price']]

#  Next sort cut and carat descending and price ascending all same time
diamonds4_sorted = diamonds4.sort_values(['cut','carat','price'],ascending=[False,False,True])

diamonds4_sorted.head(10)

cut  carat  price
26444  Very Good   4.00  15984
26100  Very Good   3.04  15354
16283  Very Good   3.00   6512
27024  Very Good   2.74  17164
27065  Very Good   2.74  17184
25525  Very Good   2.70  14341
21566  Very Good   2.68   9665
22413  Very Good   2.63  10437
25778  Very Good   2.58  14749
23960  Very Good   2.54  12095

In [117]:
# next we drop duplicates and see the cheapest diamonds per cut and carat groups
diamonds4_sorted.drop_duplicates(subset=['cut','carat']).head(10)

cut  carat  price
26444  Very Good   4.00  15984
26100  Very Good   3.04  15354
16283  Very Good   3.00   6512
27024  Very Good   2.74  17164
25525  Very Good   2.70  14341
21566  Very Good   2.68   9665
22413  Very Good   2.63  10437
25778  Very Good   2.58  14749
23960  Very Good   2.54  12095
27265  Very Good   2.52  17689

By default, drop_duplicates keeps the very first appearance, but this behavior may be modified by passing the keep parameter last to select the last row of each group or False to drop all duplicates entirely.

# Duplicating nlargest with sort_values

The previous two recipes work similarly by sorting values in slightly different manners. Finding the top n values of a column of data is equivalent to sorting the entire column descending and taking the first n values. Pandas has many operations that are capable of doing this in a variety of ways.

Getting ready
In this recipe, we will replicate the Selecting the smallest from the largest recipe with the sort_values method and explore the differences between the two.

How to do it...
Use sort_values to replicate the first part of the expression and grab the first 100 rows with the head method:


In [118]:
diamonds5 = diamonds[['cut', 'carat', 'price']]
diamonds5.sort_values('carat',ascending=False).head(100).head()

cut  carat  price
27415     Fair   5.01  18018
27630     Fair   4.50  18531
27130     Fair   4.13  17329
25999  Premium   4.01  15223
25998  Premium   4.01  15223

In [119]:
diamonds5.sort_values('carat',ascending=False).head(100).head().sort_values('price').head()

cut  carat  price
25999  Premium   4.01  15223
25998  Premium   4.01  15223
27130     Fair   4.13  17329
27415     Fair   5.01  18018
27630     Fair   4.50  18531

In [120]:
diamonds5.nlargest(100, 'carat').tail()

cut  carat  price
26818    Ideal   2.53  16709
26915  Premium   2.53  16934
26999     Fair   2.53  17103
27227  Premium   2.53  17591
27522  Premium   2.53  18254

In [121]:
diamonds5.sort_values('carat', ascending=False).head(100).tail()

cut  carat  price
26999     Fair   2.53  17103
25964  Premium   2.53  15148
26455    Ideal   2.53  15993
27522  Premium   2.53  18254
26818    Ideal   2.53  16709

How it works...
The sort_valuesmethod can nearly replicate nlargest by chaining the head method after the operation, as seen in step 2. Step 3 replicates nsmallest by chaining another sort_values and completes the query by taking just the first five rows with the head method.

Take a look at the output from the first DataFrame from step 1 and compare it with the output from step 3. Are they the same? No! What happened? To understand why the two results are not equivalent, let's look at the tail of the intermediate steps of each recipe:

The issue arises because more than 100 movies exist with a rating of at least 8.4. Each of the methods, nlargest and sort_values, breaks ties differently, which results in a slightly different 100-row DataFrame. 


There's more...
If you look at the nlargest documentation, you will see that the keep parameter has three possible values, first, last, and False. From my knowledge of other pandas methods, keep=False should allow all ties to remain part of the result. Unfortunately, pandas raises an error when attempting to do this. I created an issue with pandas development team on GitHub to make this enhancement (http://bit.ly/2fGrCMa).

# Calculating a trailing stop order price

In [0]:
import pandas_datareader as pdr

In [123]:
tsla = pdr.DataReader('tsla', data_source='google',start='2017-1-1')
tsla.head(8)

ImmediateDeprecationError: ignored

In [0]:
tsla_close = tsla['Close']

In [0]:
tsla_cummax = tsla_close.cummax()
tsla_cummax.head(8)

Date
2017-01-03    216.99
2017-01-04    226.99
2017-01-05    226.99
2017-01-06    229.01
2017-01-09    231.28
2017-01-10    231.28
2017-01-11    231.28
2017-01-12    231.28
Name: Close, dtype: float64

In [0]:
tsla_trailing_stop = tsla_cummax * .9
tsla_trailing_stop.head(8)

Date
2017-01-03    195.291
2017-01-04    204.291
2017-01-05    204.291
2017-01-06    206.109
2017-01-09    208.152
2017-01-10    208.152
2017-01-11    208.152
2017-01-12    208.152
Name: Close, dtype: float64

## There's more...

In [0]:
def set_trailing_loss(symbol, purchase_date, perc):
    close = pdr.DataReader(symbol, 'google', start=purchase_date)['Close']
    return close.cummax() * perc

In [0]:
msft_trailing_stop = set_trailing_loss('msft', '2017-6-1', .85)
msft_trailing_stop.head()

Date
2017-06-01    59.585
2017-06-02    60.996
2017-06-05    61.438
2017-06-06    61.642
2017-06-07    61.642
Name: Close, dtype: float64